scrapd/core/apd.py

```diff
@@ -534,6 +534,27 @@ def parse_crashes_field(page):
     return match_pattern(page, crashes_pattern)
 
 
+def parse_time_field(page):
+    """
+    Extract the time from the content of the fatality page.
+
+    :param str page: the content of the fatality page
+    :return: a string representing the case number.
+    :rtype: str
+    """
+    time_pattern = re.compile(
+        r'''
+        Time:           # The name of the desired field.
+        .*?             # Any character (lazy).
+        (\d{1,2}:\d{2}  # Time format (##:## or #:##).
```

Comment:
Thinking about it a bit more about it, we can even further improve by restricting the time format to something which matches exactly a valid time. This would prevent us from having invalid dates in the data sets. We want to make sure that we do not accept times like `35:73`.

Something like that may do the trick`((?:[01]?\d):[0-5]\d)` (I have not tested it!).

Reply:
This one was tested on regex101.com (<https://regex101.com/r/I2VCNa/2>) and seems to work fine: `Time:.*?(:?\s+|\D)((?:[01]?\d):[0-5]\d)`

Reply:
Ok, this is a bit long-winded. You may want to make some tea and get comfortable. I apologize in advance. 😜 

https://regex101.com/r/I2VCNa/5
`r'Time:(?:\s+|\D*?)([012]?\d:[0-5]\d\s+[AaPp]?\.?[Mm]?\.?)'`

This regex is pretty robust, allowing for multiple formats of am/pm (upper- and lowercase, with and without periods, as well as no am/pm at all). However it still allows 24-29 to slip through in the 'hour' location. It seems unlikely that we'll encounter those entries, but if we want to account for them, perhaps we can either:
- Have a longer, more specific regex, or
- Validate the time format inside `parse_time_field()` with an auxiliary function

For instance, this function will pull out the digits (either #:## or ##:##) in a string that may or may not include am/pm (in any of it's variations), and check if they are a valid time.

```python
def valid_time(time, timeformat="%H:%M"):
    regex = re.compile(r'\d{1,2}:\d{2}')
    time = re.search(regex, time).group()
    try:
        datetime.datetime.strptime(time, timeformat)
        return True
    except ValueError:
        return False
```

...so I imagine we could go further and have `parse_time_field()` clean-up/return the desired time format that will be used later on down the pipeline (in the dashboard?), since we can't be certain that APD will be consistent in their am/pm formatting.

Also,
- I don't fully understand how to modify the `parse_page_content()` function to include `parse_time_field()`. Would I remove the `Fields.TIME` entry from the `searches` list (L467), and then mimic the code under `# Parse the Case field` (L475)?
- Hypothetically, if `valid_time()` were to be incorporated, would it go in `date_utils.py`?
- How do I go about making any of these changes to the PR, while making sure my branch is up to date with any changes that may have happened recently?

I know you're a busy guy, but if this would be easier to discuss in person, I can meet up with you whenever. 😄

Reply:
Your approach is very good! However I think what we did wrong was to try to validate the 12h-format and 24h-format all together in one regex.  Because it will also validate `Time: 18:46 pm` which is completely wrong (<https://regex101.com/r/I2VCNa/6>).

I think we can solve this issue by having 2 regexes. Try the 12h-format first, if it catches, return the match, other try the 24h-format. 

Interesting links I found after Googling for 1 minute:
* [12h-format regex](https://stackoverflow.com/questions/35667601/how-to-validate-12-hour-time-with-regex-regular-expression)
* [24h-format regex](https://stackoverflow.com/questions/1494671/regular-expression-for-matching-time-in-military-24-hour-format)
* [O'reilly cookbook](https://www.oreilly.com/library/view/regular-expressions-cookbook/9781449327453/ch04s06.html)

To answer your other questions:
> I don't fully understand how to modify the parse_page_content() function to include parse_time_field(). Would I remove the Fields.TIME entry from the searches list (L467), and then mimic the code under # Parse the Case field (L475)?

Yes, you are absolutely correct, that's what you must do. Just mimic the `Fields.CRASHES` instead (l481) as we allow time to be empty.

> Hypothetically, if valid_time() were to be incorporated, would it go in date_utils.py?

Correct again, but hopefully, with the 2 regexes approach we won't need it.

> How do I go about making any of these changes to the PR, while making sure my branch is up to date with any changes that may have happened recently?

Just keep committing and pushing to your branch. If it gets out of sync, you can rebase your branch on top of master. I wrote a guide here to help you with this task: https://request-yo-racks.github.io/docs/guides/update-branch/. You can also click on the `Update branch` button GitHub shows you and pull the update back to your branch with `git pull origin <your_branch_name>`